File: solutions/507/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Fri Dec 31 09:52:59 2021
> Description:   
 ************************************************************************/
#include <cmath> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    bool checkPerfectNumber(int num) {
        // check input
        if (num == 1) return false;
        int total = 0;
        for (int i = 1; i < int(sqrt(num))+1; ++i) {
            if (num % i == 0) {
                total += i;
                if (i != 1 && i != sqrt(num))
                    total += num / i;
            }
            if (total > num) return false;
        }
        return total == num;
    }
};

TEST(Test507, SimpleTest) {
    int num = 28; 
    Solution s; 
    EXPECT_TRUE(s.checkPerfectNumber(num));
    num = 7; 
    EXPECT_FALSE(s.checkPerfectNumber(num));
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/507/main.py
import math 

class Solution:
    def checkPerfectNumber(self, num: int) -> bool:
        if num == 1: 
            return False 
        total = 0 
        for i in range(1, math.floor(math.sqrt(num))+1): 
            if num % i == 0: 
                total += i 
                if i != 1 and i != math.sqrt(num): 
                    total += num / i 
            if total > num: 
                return False 
        return total == num 

if __name__ == "__main__": 
    s = Solution()
    num = 28 
    assert s.checkPerfectNumber(num)
    num = 7 
    assert not s.checkPerfectNumber(num)
